### PR TITLE
Ayomide/testsuit

### PR DESCRIPTION
**Pull request by Adebara123**

### CODE REVIEW

File: modules/ismp/testsuite/src/lib.rs [hunk 1]
```diff
@@ -389,20 +396,349 @@ where
 
 /// This should prevent a request from timing out on a proxy when there exists a consensus client
 /// for the request destination
-pub fn prevent_request_timeout_on_proxy_with_known_state_machine() {}
+/// The State machine for the proxy is assumed in this test to be ``StateMachine::Ethereum(Ethereum::ExecutionLayer)``
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:399]: This documentation is wrong, the proxy state machine is Kusam(2000).

File: modules/ismp/testsuite/src/lib.rs [hunk 2]
```diff
@@ -389,20 +396,349 @@ where
...
+/// the State machine for the host is assumed in this test to be ``StateMachine::Polkadot(1000)``
+/// The destination state machine for the test is assumed to be ``StateMachine::Kusama(1000)``
+pub fn prevent_request_timeout_on_proxy_with_known_state_machine<H, D>(host: &H, dispatcher: &D)  -> Result<(), &'static str>
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:402]: Allow the consumer of the test specify what the proxy_state_machine is and what the directly connected chain is for the purpose of the test.

File: modules/ismp/testsuite/src/lib.rs [hunk 3]
```diff
@@ -389,20 +396,349 @@ where
...
+
+let consensus_clients= host.consensus_clients();
+    assert!(consensus_clients.len() > 1);
+
+// assert that destination chain concensus client is in the Host list of clients 
+// destination chain concensus in this test is assumed to be MOCK_CONSENSUS_CLIENT_ID
+
+
+let proxy_consensus_client_id = consensus_clients.iter().find(|client| client.state_machine(proxy_state_machine).ok().is_some()).expect("The proxy consensus client should be set for this test").consensus_client_id();
+let destination_consensus_client_id = consensus_clients.iter().find(|client| client.state_machine(direct_conn_state_machine).ok().is_some()).expect("The proxy destination chain's consensus client should be set for this test").consensus_client_id();
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:454]:
```suggestion
let destination_consensus_client_id = consensus_clients.iter().find(|client| client.state_machine(direct_conn_state_machine).ok().is_some()).expect("The directly connected chain's consensus client should be set for this test").consensus_client_id();
```

File: modules/ismp/testsuite/src/lib.rs [hunk 4]
```diff
@@ -389,20 +396,349 @@ where
...
+
+
+// For our test case we assert that there exists distinct consensus clients for the proxy and the direct connection
+
+assert_ne!(proxy_consensus_client_id, destination_consensus_client_id);
+
+let  destiation_client_= &consensus_clients[0].consensus_client_id();
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:461]: Why is this still here?

File: modules/ismp/testsuite/src/lib.rs [hunk 5]
```diff
@@ -389,20 +396,349 @@ where
...
+
+let dispatch_post = DispatchPost {
+    dest: intermediate_state.height.id.state_id,
+    from: vec![0u8; 32],
+    to: vec![0u8; 32],
+    timeout_timestamp: intermediate_state.commitment.timestamp,
+    data: vec![0u8; 64],
+    gas_limit: 0,
+};
+let post = Post {
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:475]: This post is not correct for the purposes of this test

File: modules/ismp/testsuite/src/lib.rs [hunk 6]
```diff
@@ -389,20 +396,349 @@ where
...
+
+
+
+ // Timeout message handling check for source and destination chain
+ let timeout_message = Message::Timeout(TimeoutMessage::Post {
+    requests: vec![request.clone()],
+    timeout_proof: Proof { height: intermediate_state.height, proof: vec![] },
+});
+
+handle_incoming_message(host, timeout_message).unwrap();
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:502]: Why are there two timeout checks?

**Review by Wizdave97 — CHANGES_REQUESTED** [review]

**Comment by Adebara123** [comment]
@Wizdave97 All requested changes have been resolved. Please look into it.

File: modules/ismp/testsuite/src/lib.rs [hunk 7]
```diff
@@ -19,12 +19,14 @@ pub mod mocks;
 #[cfg(test)]
 mod tests;
 
-use crate::mocks::MOCK_CONSENSUS_CLIENT_ID;
+use std::process::id;
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:22]:
```suggestion

```

File: modules/ismp/testsuite/src/lib.rs [hunk 8]
```diff
@@ -40,11 +42,15 @@ fn mock_consensus_state_id() -> ConsensusStateId {
...
+fn mock_proxy_consensus_state_id() -> ConsensusStateId {
+    *b"prox"
+}
+
 fn setup_mock_client<H: IsmpHost>(host: &H) -> IntermediateState {
     let intermediate_state = IntermediateState {
         height: StateMachineHeight {
             id: StateMachineId {
-                state_id: StateMachine::Ethereum(Ethereum::ExecutionLayer),
+                state_id: StateMachine::Bsc,
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:53]: revert this
```suggestion
                state_id: StateMachine::Bsc,
```

File: modules/ismp/testsuite/src/mocks.rs [hunk 9]
```diff
@@ -24,8 +24,11 @@ use std::{
 
 #[derive(Default)]
 pub struct MockClient;
+#[derive(Default)]
+pub struct MockProxyClient;
 
 pub const MOCK_CONSENSUS_CLIENT_ID: [u8; 4] = [1u8; 4];
+pub const MOCK_CONSENSUS_CLIENT_ID_2: [u8; 4] = [2u8; 4];
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/mocks.rs:31]:
```suggestion
pub const MOCK_PROXY_CONSENSUS_CLIENT_ID: [u8; 4] = [2u8; 4];
```

File: modules/ismp/testsuite/src/lib.rs [hunk 10]
```diff
@@ -389,20 +422,289 @@ where
...
+    proxy_state_machine: StateMachine,
+    direct_conn_state_machine: StateMachine,
+) -> Result<(), &'static str>
+where
+    H: IsmpHost,
+    D: IsmpDispatcher,
+    D::Account: From<[u8; 32]>,
+    D::Balance: From<u32>,
+{
+    let intermediate_state = setup_mock_client(host);
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:440]: Only the proxy needs to be setup for this test.
```suggestion
    let intermediate_state = setup_mock_client(host);
```

File: modules/ismp/testsuite/src/lib.rs [hunk 11]
```diff
@@ -389,20 +422,289 @@ where
...
+        .unwrap();
+
+    let timeout_message = Message::Timeout(TimeoutMessage::Post {
+        requests: vec![request.clone()],
+        timeout_proof: Proof { height: proxy.height, proof: vec![] },
+    });
+
+    let res = handle_incoming_message(host, timeout_message).unwrap();
+
+    // Assert that the dispatch results are empty
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:510]:
```suggestion
    // Assert that the dispatch results are empty which means the timeout message was ignored.
```

**Review by Wizdave97 — APPROVED** [review]

**Comment by Wizdave97** [comment]
Update the branch

**Comment by Adebara123** [comment]
@Wizdave97  Branch updated

**Pull request merged**